internal/fields/mappings.go

```diff
@@ -624,22 +687,295 @@ func (v *MappingValidator) validateMappingsNotInPreview(currentPath string, chil
 			continue
 		}
 
-		// TODO: validate mapping with dynamic templates first than validating with ECS
-		// just raise an error if both validation processes fail
+		// validate whether or not the field has a corresponding dynamic template
+		if len(dynamicTemplates) > 0 {
+			err := v.matchingWithDynamicTemplates(fieldPath, def, dynamicTemplates)
+			if err == nil {
+				continue
+			}
+		}
 
-		// are all fields under this key defined in ECS?
+		// validate whether or not all fields under this key are defined in ECS
 		err = v.validateMappingInECSSchema(fieldPath, def)
 		if err != nil {
-			logger.Warnf("undefined path %q (pending to check dynamic templates)", fieldPath)
 			errs = append(errs, fmt.Errorf("field %q is undefined: %w", fieldPath, err))
 		}
 	}
 	return errs.Unique()
 }
 
+// matchingWithDynamicTemplates validates a given definition (currentPath) with a set of dynamic templates.
+// The dynamic templates parameters are based on https://www.elastic.co/guide/en/elasticsearch/reference/8.17/dynamic-templates.html
+func (v *MappingValidator) matchingWithDynamicTemplates(currentPath string, definition map[string]any, dynamicTemplates []map[string]any) error {
+
+	parseSetting := func(value any) ([]string, error) {
+		all := []string{}
+		switch v := value.(type) {
+		case []any:
+			for _, elem := range v {
+				s, ok := elem.(string)
+				if !ok {
+					return nil, fmt.Errorf("failed to cast to string: %s", elem)
+				}
+				all = append(all, s)
+			}
+		case any:
+			s, ok := v.(string)
+			if !ok {
+				return nil, fmt.Errorf("failed to cast to string: %s", v)
+			}
+			all = append(all, s)
+		default:
+			return nil, fmt.Errorf("unexpected type for setting: %T", value)
+
+		}
+		return all, nil
+	}
+
+	stringMatchesPatterns := func(regexes []string, elem string) (bool, error) {
+		applies := false
+		for _, v := range regexes {
+			if !strings.Contains(v, "*") {
+				// not a regex
+				continue
+			}
+
+			match, err := regexp.MatchString(v, elem)
+			if err != nil {
+				return false, fmt.Errorf("failed to build regex %s: %w", v, err)
+			}
+			if match {
+				applies = true
+				break
+			}
+		}
+		return applies, nil
+	}
+
+	stringMatchesWildcards := func(regexes []string, elem string) (bool, error) {
+		for i, v := range regexes {
+			r := strings.ReplaceAll(v, ".", "\\.")
+			r = strings.ReplaceAll(r, "*", ".*")
+			// Force to match beginning and ending
+			r = fmt.Sprintf("^%s$", r)
+
+			regexes[i] = r
+		}
+		return stringMatchesPatterns(regexes, elem)
+	}
+
+	fieldType := mappingParameter("type", definition)
+	if fieldType == "" {
+		return fmt.Errorf("missing type parameter for field: %q", currentPath)
+	}
+
+	for _, template := range dynamicTemplates {
```

Comment:
Nit. This loop mixes parsing of the dynamic templates and its execution. Consider unmarshalling the dynamic templates first, so this loop becomes something like:
```
for _, template := dynamicTemplates {
    matches, err := template.Matches(currentPath, definition)
    if err != nil {
        return err
    }
    if matches {
        return nil
    }
}
```
The parsing of the dynamic templates could be done once, in `validateMappingsNotInPreview`.